pydantic/json_schema.py

```diff
@@ -0,0 +1,1081 @@
+from enum import Enum
+from ipaddress import IPv4Address, IPv4Interface, IPv4Network, IPv6Address, IPv6Interface, IPv6Network
+from pathlib import PurePath
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Counter,
+    Dict,
+    List,
+    NewType,
+    Optional,
+    Pattern,
+    Sequence,
+    Type,
+    Union,
+    cast,
+)
+from uuid import UUID
+
+from pydantic_core import CoreSchema, CoreSchemaType, core_schema
+from pydantic_core.core_schema import TypedDictField
+from typing_extensions import Literal, TypedDict, TypeGuard
+
+from ._internal._core_metadata import CoreMetadataHandler
+from ._internal._typing_extra import all_literal_values, is_namedtuple
+from ._internal._utils import get_model, lenient_issubclass
+from .errors import PydanticInvalidForJsonSchema, PydanticUserError
+from .networks import IPvAnyAddress, IPvAnyInterface, IPvAnyNetwork
+
+if TYPE_CHECKING:
+    from .dataclasses import Dataclass
+    from .main import BaseModel
+
+JsonSchemaValue = Dict[str, Any]
+
+# ##### JSON Schema Metadata Manipulation #####
+# Keys missing in a JsonSchemaMetadata should be treated the same as they would be if present with a value of None.
+# This is important to ensure that it is possible to "remove" an override by setting it to None.
+JsonSchemaMetadata = TypedDict(
+    'JsonSchemaMetadata',
+    {
+        # ### "Pre-processing" of the JSON schema
+        # If not None, this CoreSchema will be used to generate the JSON schema instead of the "real one"
+        # You can use a callable to defer evaluation of the CoreSchema until it's needed
+        'core_schema_override': Union[CoreSchema, Callable[[], CoreSchema], None],
+        # A reference to the source class if appropriate; useful when working with some of the plain function schemas
+        'source_class': Union[Type[Any], None],
+        # ### "Miscellaneous properties" that are available for all JSON types
+        # (see https://json-schema.org/understanding-json-schema/reference/generic.html)
+        'title': Optional[str],
+        'description': Optional[str],
+        'examples': Optional[List[Any]],
+        'deprecated': Optional[bool],
+        'readOnly': Optional[bool],
+        'writeOnly': Optional[bool],
+        '$comment': Optional[str],
+        # Note: 'default', which is included with these fields in the JSON Schema docs, is handled by CoreSchema
+        # ### "Post-processing" of the JSON schema
+        # A catch-all for arbitrary data to add to the schema
+        'extra_updates': Optional[Dict[str, Any]],
+        # A final function to apply to the JSON schema after all other modifications have been applied
+        # If you want to force specific contents in the generated schema, you can use a function that ignores the
+        # input value and just return the schema you want.
+        'modify_js_function': Optional[Callable[[JsonSchemaValue], None]],
+    },
+    total=False,
+)
+
+
+def merge_js_metadata(
```

Comment:
boring I know, but maybe we should be consistent about calling `json_schema`, `js` internally?

But not that bothered if you disagree.

Reply:
I think there are cases where we should use `json_schema` or `JsonSchema` in full form for clarity; I'm most inclined to shorten it in private APIs where someone would be more likely to know that `js` is short for `json_schema`. I'll take a look though everywhere we use it in full form and make sure I think it's good.

Reply:
agreed, ignore this one.